### ISSUE TITLE
merge(mode='dot') and merge(mode='cos') not interchangeable; NaN using custom cos merge

### ISSUE THREAD

**Issue by phdowling** [fchollet/keras#4553]
I am trying to [implement the ABCNN](https://github.com/phdowling/abcnn-keras/blob/master/abcnn.py) introduced [here](https://arxiv.org/abs/1512.05193), which entails evaluating a word-by-word similarity matrix of two input sentences. If one sentence has 5 words, and the other 8, this matrix should be 5 x 8. Applying a merge layer with mode="dot" is the right functional interface, i.e. it leads to the correct dimensions, but it's incomplete in that I actually need a similarity measure. I figured I should be able to apply the same layer with "cos" mode, but it leads to an error:
```
  File "/Users/dowling/Development/keras-abcnn/abcnn.py", line 49, in MatchScore
    dot = merge([l, r], mode="cos")
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/keras/engine/topology.py", line 1528, in merge
    name=name)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/keras/engine/topology.py", line 1186, in __init__
    node_indices, tensor_indices)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/keras/engine/topology.py", line 1221, in _arguments_validation
    'Layer shapes: %s' % input_shapes)
Exception: Only layers of same output shape can be merged using cos mode. Layer shapes: [(None, 14, 64), (None, 10, 64)]
```
This is the main point of this issue - I assumed that cos and dot should be interchangable and handle the same input dims, but that seems not to be the case. Is this desired behavior?

The second part of this issue is more a programming problem, so I'm not sure if here is the right place to ask, but perhaps this is relevant to Keras somehow:
Because the "cos" merge did not work as expected, I tried implementing cosine similarity myself (as well as the "euclidean" similarity introduced in the paper (`sim(a, b) = 1 / (1 + euclidean_distance(a, b))`)).
This seems to work when I [test the functions standalone](https://github.com/phdowling/abcnn-keras/blob/master/test_matchscore.py), but included in the my model architecture they lead to NaN values (weights and predictions) once backprop is applied. For a runnable example, see: https://github.com/phdowling/abcnn-keras/blob/master/abcnn.py. Note that with the `abcnn_1` flag set to False (attention only in upper layers), the NaN values only occur on some training runs. I _never_ got NaN values using only a `dot` merge, even though my own merges are quite similar in their implementation.

Any help is appreciated!

**Comment by patyork** [comment]
To answer part 1: the cosine similarity requires data with the same dimensionality, as you've seen. This is required by the [definition](https://en.wikipedia.org/wiki/Cosine_similarity#Definition) of cosine similarity.

As to the NaNs, this is a pretty common issue when the function includes a complex denominator that can become zero or cause an underflow/overflow. Keras functions generally build in a check, which you [can see here](https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1283) in the Keras 'cos' merge, which clips the denomitator to a small but non-zero number. There are many examples of this throughout the loss/activation functions in Keras, and I'd recommend doing the same in anything you add to Keras.

**Comment by patyork** [comment]
I'd really recommend thinking about what the cosine similarity between two matrices of shapes `[(None, 14, 64), (None, 10, 64)]` actually means, a look at the definition again, and then whether or not this makes mathematical sense.

Cosine similarity is not just a dot product, it involves a denominator of magnitudes, and this is where the dimensionality issue comes in. When applying a vector function (magnitude/norm) over a list of vectors, the behavior changes. The denominator becomes, dimensionality wise: `[None, 14, 1] * [None, 10, 1]`; not a single number for magnitude, but a list/vector of magnitudes that must be multiplied together. I recommend yo do a quick sanity check of what the shape of the terms in the denominator should be, and then an attempt at a multiplication of those two shapes; this would be an example: `np.ones((14,1)) * np.ones((10, 1))` -> a numpy shape error.

**Comment by MaratZakirov** [comment]
I must say currently I also have NaN with 'cos' merge (keras 1.1.0) and strange shape after cos merge (which is different from merge dot shape)
